week3_robot_solved

insungg opened the pull request:
week3_robot_solved

Commit:
week3_robot_solved

Commit:
revised getInt() by using try catch

accforaus commented on inseung/week3_class_Robot/robot.kt:
```diff
@@ -0,0 +1,31 @@
+
+/*
+
+abstract class Robot 에서 멤버인 변수들을 초기화 하지 않고 선언하려면 어떻게 하면 좋을까요 일단 abstract 형으로 했습니다
```
`primitive` 값들, `Int`, `Long`, `Float`, `Double` 값들을 제외하고는 `lateinit` 키워드 사용하면 하면돼

```kotlin
lateinit var model: String
lateinit var position: Coordinate2D
```

```kotlin

abstract class Robot {
    lateinit var model: String
    var distance: Int = 0
    var price: Int = 0
    var ID = 0
    
    open fun show() {}
}

class CleaningRobot : Robot() {
    var cleaningPower: Int = 10
    
    fun cleaningStart() {}
    fun cleaningStop() {}
}

class DogRobot : Robot() {
    var barkPower: Int = 10
    
    fun barkStart() {}
    fun barkStop() {}
}
```

이렇게 클래스 구현하면 되겟넹

insungg replied:
robot의 멤버 distance, price, ID 같은 것들은 선언시에 초기화해야되서 var 형으로만 선언할 수 밖에 없나요?
lateinit 으로 생성한 model 과 position을 상속받은 CleanerRobot과 DogRobot에서 초기화 할 수는 없나요?

accforaus replied:
초기화 하고 싶으면 생성자 만들면 되겟넹

```kotlin
abstract Robot(var model: String = "", ... )
```

insungg replied:
앗...그렇군요...간다하네요
혹시 객체의 type을 반환하는 연산자?? 같은것도 있나요?

accforaus replied:
있는데.. 많이 복잡한데;; 

그냥 일단은 아직 `reflection` 모르니까 `is`로만 사용하는걸로..

accforaus commented on inseung/week3_class_Robot/robot.kt:
```diff
@@ -0,0 +1,31 @@
+abstract class Robot 에서 멤버인 변수들을 초기화 하지 않고 선언하려면 어떻게 하면 좋을까요 일단 abstract 형으로 했습니다
+
+Robot의 자식들인 CleaningRobot 과 DogRobot 에서 Robot 의 멤버를 쓰려면 전부 override해서 초기화 해야되는데 너무 불편합니다. 원래 이런건가요?
+
```
`abstract` 키워드 사용하면 무조건 `override` 키워드 사용해야 하니까 일반 변수로 선언 해주고, 자식 클래스는 그냥 접근해서 사용하면 돼

위에 참고.. ㅋㅋ

accforaus commented on inseung/week3_class_Robot/InputHelper.kt:
```diff
@@ -0,0 +1,20 @@
+    private lateinit var scanner: Scanner
+
+    fun getInt() : Int {
+        scanner = Scanner(System.`in`)
```
다른 값을 입력했을 시 오류처리 안했넹

```kotlin
fun getInt() : Int {
    while (true) {
        try {
            return Scanner(System.`in`).nextInt()
        } catch (e: Exception) {
            println("정수 값만 입력 가능합니다.")
        }
    }
}
```

accforaus commented on inseung/week3_class_Robot/robot.kt:
```diff
@@ -0,0 +1,31 @@
+
+SortWith 에 인자로 주는 Comparator 는 함수 or class 와 같은 object 둘 다 가능한가요? 어떤 타입을 줘야하나요
+
+C 언어에서 #define 역할을 하는 것이 Kotlin 에 있나요 ... mapSize인 200을 #define mapSize 200 이렇게 쓰고 싶네요
```
```kotlin
// constants.kt
const val MAP_SIZE = 200
```
이렇게 정의 해주고 다른 파일에서 사용해주면 됩

accforaus commented on inseung/week3_class_Robot/Controller.kt:
```diff
@@ -0,0 +1,211 @@
+                tempC.position.y = InputHelper.getInt()
+                tempC.price = InputHelper.getInt()
+                tempC.cleaningPower = InputHelper.getInt()
+                collection.add(tempC)
```
```kotlin
print ("세부 사항을 입력하세요(ID, 로봇명, x, y, 가격, cleaningPower): ")
collection.add(CleaningRobot ().apply {
        ID = InputHelper.getInt()
        model = InputHelper.getWord()
        position.x = InputHelper.getInt()
        position.y = InputHelper.getInt()
        price = InputHelper.getInt()
        cleaningPower = InputHelper.getInt()
    })
```
이렇게 간단하게 쓸수 있어

insungg replied:
왜 apply 문 안쪽에서 this.ID 이런식으로 접근하지 않아도 되나요?

accforaus replied:
`apply`문 에서는 자동추론이 가능해서, 중복되는 이름이 없는한 `this`키워드는 생략해도 돼 ㅋㅋ

만약 `let`을 쓴다면 `it.ID` 이렇게 접근해야 되는데

accforaus commented on inseung/week3_class_Robot/robot.kt:
```diff
@@ -0,0 +1,31 @@
+
+Robot의 자식들인 CleaningRobot 과 DogRobot 에서 Robot 의 멤버를 쓰려면 전부 override해서 초기화 해야되는데 너무 불편합니다. 원래 이런건가요?
+
+SortWith 에 인자로 주는 Comparator 는 함수 or class 와 같은 object 둘 다 가능한가요? 어떤 타입을 줘야하나요
```
```kotlin

class RobotNameComparator : Comparator<Robot> {
    override fun compare(o1: Robot?, o2: Robot?): Int {
        return o1!!.model.compareTo(o2!!.model)
    }
}

class RobotIdComparator : Comparator<Robot> {
    override fun compare(o1: Robot?, o2: Robot?): Int {
        return o1!!.ID.compareTo(o2!!.ID)
    }
}

class RobotPriceComparator : Comparator<Robot> {
    override fun compare(o1: Robot?, o2: Robot?): Int {
        return o1!!.price.compareTo(o2!!.price)
    }
}

list.sortWith(RobotPriceComparator())
list.sortWith(RobotNameComparator())
list.sortWith(RobotIdComparator())
```

이렇게 정의하고 사용하면 돼

accforaus commented on inseung/week3_class_Robot/Controller.kt:
```diff
@@ -0,0 +1,211 @@
+        collection.sortWith(Comparator { data1, data2 -> data1.price - data2.price } )
+        collection.forEachIndexed { index, robot -> print("${index + 1}    "); robot.show() }
+    }
+
```
구현 잘했넹 중복 코드가 많으니까 줄여주면 되겠다

Commit:
Week3 Robot Revised

Make Package to Robot Class

Change Robot List sorting Comparator to Comparators package and classes

Remove abstract declared members of Robot Class

Add MAP_SIZE constant

Add Input Exception in x, y coordinates and int values for selecting menu options

Commit:
Week3 Robot Revised

Make Package to Robot Class

Remove abstract declared members of Robot Class

Change Robot List sorting Comparator to Comparators package and classes

Add MAP_SIZE constant

Add Input Exception in x, y coordinates and int values for selecting menu options

accforaus commented on inseung/week3_class_Robot/korea/Hwang/ccp/controller/Controller.kt:
```diff
@@ -0,0 +1,225 @@
+        collection.sortWith(RobotPriceComparator())
+        collection.forEachIndexed { index, thisObj -> print("${index + 1}    "); thisObj.show() }
+    }
+
```
`printListBy...` 요부분 전부 중복되니까 한번 줄여보자

accforaus commented on inseung/week3_class_Robot/korea/Hwang/ccp/controller/Controller.kt:
```diff
@@ -0,0 +1,225 @@
+            else { println("1부터 3사이의 정수를 입력해 주세요."); continue }
+        }
+    }
+
```
`if else`문도 전부 `when`문으로 바꾸면 깔꼼 해질거야

accforaus commented on inseung/week3_class_Robot/korea/Hwang/ccp/utils/RobotComparators/RobotIDComparator.kt:
```diff
@@ -0,0 +1,9 @@
+class RobotIDComparator : Comparator<Robot> {
+    override fun compare(o1: Robot?, o2: Robot?): Int {
+        return o1!!.ID.compareTo(o2!!.ID)
+    }
```
```kotlin
class RobotIDComparator : Comparator<Robot> {
	    override fun compare(o1: Robot?, o2: Robot?): Int = o1!!.ID.compareTo(o2!!.ID)
}
```

깔끔..

accforaus commented on inseung/week3_class_Robot/korea/Hwang/ccp/utils/RobotComparators/RobotModelComparator.kt:
```diff
@@ -0,0 +1,9 @@
+package korea.Hwang.ccp.utils.RobotComparators
+
+import korea.Hwang.ccp.domain.Robot
+
```
`RobotIDComparator` 참고

accforaus commented on inseung/week3_class_Robot/korea/Hwang/ccp/utils/RobotComparators/RobotPriceComparator.kt:
```diff
@@ -0,0 +1,9 @@
+package korea.Hwang.ccp.utils.RobotComparators
+
+import korea.Hwang.ccp.domain.Robot
+
```
`RobotIDComparator` 참고

accforaus commented on inseung/week3_class_Robot/korea/Hwang/ccp/utils/RobotComparators/RobotTypeComparator.kt:
```diff
@@ -0,0 +1,10 @@
+
+import korea.Hwang.ccp.domain.CleaningRobot
+import korea.Hwang.ccp.domain.Robot
+
```
`RobotIDComparator` 참고

accforaus commented on inseung/week3_class_Robot/main.kt:
```diff
@@ -0,0 +1,8 @@
+import korea.Hwang.ccp.controller.Controller
+
+const val MAP_SIZE = 200
+
```
`Controller` 클래스가 굳이 `class`일 필요가 있을까? `object` 형일때 장점은?

insungg replied:
음 controller 는 어차피 run method 만 있으면 되니까 main 에서 굳이 메모리에 올릴 필요 없이 바로 인스턴스 할 수 있는게 좋지 않을까요
혹시... 로봇 자식이 100개가 어디서 생기는지 알 수 있습니까

accforaus replied:
로봇 자식 100개는 사용자가 만드는거야 ㅋㅋ 상속받아서

accforaus commented on inseung/week3_class_Robot/korea/Hwang/ccp/utils/Coordinate2D.kt:
```diff
@@ -0,0 +1,24 @@
+        fun moveX (direction: Int)  {
+            if (x + distance !in 0..MAP_SIZE) return //이동후 위치가 map의 크기 200과 0을 넘어가면 움직이지 않음
+            else x = x + direction
+        }
```
```kotlin
               fun moveX (direction: Int)  {
	            if (x + distance in 0..MAP_SIZE) x = x + direction
	        }
```

Commit:
Change Controller's type class to object

Commit:
Revise comparator's return part into more simple form

Commit:
Revise moveX, moveY return part into more simple form

Commit:
Change instantiate of Controller's run method
